### PR TITLE
Refactor bag solver

### PR DESCRIPTION
**Pull request by mudathirmahgoub**

### CODE REVIEW

File: src/theory/bags/solver_state.cpp [hunk 1]
```diff
@@ -132,6 +132,20 @@ void SolverState::collectDisequalBagTerms()
...
 }
 
+void SolverState::registerCountSkolem(Node n, Node e, Node skolem)
+{
+  Node nRep = getRepresentative(n);
+  Node eRep = getRepresentative(e);
+  d_countSkolems[n][e] = skolem;
+}
+
+Node SolverState::getCountSkolem(Node n, Node e)
```
Comment (ajreynol) [inline src/theory/bags/solver_state.cpp:142]: Can you make this a `const` method and ensure that the entry in the map exists?

Comment (mudathirmahgoub) [reply src/theory/bags/solver_state.cpp]: replaced with `getElementCountPairs`

File: src/theory/bags/solver_state.h [hunk 2]
```diff
@@ -58,20 +59,25 @@ class SolverState : public TheoryState
    * (assert (= 0 (bag.count x B)))
    * element x is associated with bag B, albeit x is definitely not in B.
    */
-  const std::set<Node>& getElements(Node B);
+  std::set<Node> getElements(Node B);
   /** initialize bag and count terms */
-  void initialize();
+  std::vector<Node> initialize();
```
Comment (ajreynol) [inline src/theory/bags/solver_state.h:64]: Document the return value here.

File: src/theory/bags/solver_state.h [hunk 3]
```diff
@@ -84,7 +90,7 @@ class SolverState : public TheoryState
   /** collection of bag representatives */
   std::set<Node> d_bags;
   /** bag -> associated elements */
-  std::map<Node, std::set<Node>> d_bagElements;
+  std::map<Node, std::vector<std::pair<Node, Node>>> d_bagElements;
```
Comment (ajreynol) [inline src/theory/bags/solver_state.h:93]: Can you document the pair here?

Comment (ajreynol) [reply src/theory/bags/solver_state.h:93]: Also, this seems to be a fairly important data structure, can you document a bit more the lifetime of this?  Is this cleared on full effort? When is it valid?

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/theory/bags/solver_state.h (hunk 2 above)
Comment (mudathirmahgoub) [reply src/theory/bags/solver_state.h:67]: Done.

**Review by ajreynol — APPROVED** [review]

**Pull request merged**